pkg/storage/s3/s3.go

```diff
@@ -0,0 +1,412 @@
+func (s *Store) FindProfiles(ctx context.Context, params *storage.FindProfilesParams) ([]profile.Meta, error) {
+	return s.find(ctx, params)
+}
+
+// FindProfileIDs calls FindProfiles and returns just the IDs.
+func (s *Store) FindProfileIDs(ctx context.Context, params *storage.FindProfilesParams) ([]profile.ID, error) {
+	metas, err := s.find(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+	ids := make([]profile.ID, len(metas))
+	for i := range metas {
+		ids[i] = metas[i].ProfileID
+	}
+	return ids, nil
+}
+
+func (s *Store) find(ctx context.Context, params *storage.FindProfilesParams) ([]profile.Meta, error) {
+	if params.Service == "" {
+		return nil, fmt.Errorf("empty service")
+	}
+
+	if params.CreatedAtMin.IsZero() {
+		return nil, fmt.Errorf("empty created_at")
+	}
+
+	createdAtMax := params.CreatedAtMax
+	if createdAtMax.IsZero() {
+		createdAtMax = time.Now().UTC()
+	}
+
+	// TODO: is this max ok?
+	limit := params.Limit
+	if limit == 0 {
+		limit = 100
+	}
+
+	input := &s3.ListObjectsV2Input{
+		Bucket:     &s.S3Bucket,
+		Prefix:     aws.String(prefix(params)),
+		StartAfter: aws.String(startAfter(params)),
+		MaxKeys:    aws.Int64(1000),
+	}
+
+	metas := []profile.Meta{}
+	err := s.svc.ListObjectsV2PagesWithContext(ctx, input,
+		func(page *s3.ListObjectsV2Output, _ bool) bool {
+			for _, object := range page.Contents {
+				if object.Key == nil {
+					continue
+				}
+				m, err := meta(*object.Key)
+				if err != nil {
+					s.logger.Error(err)
+					continue
+				}
+
+				if m.CreatedAt.After(createdAtMax) {
+					return false
+				}
+
+				if !includes(m.Labels, params.Labels) {
+					continue
+				}
+				metas = append(metas, *m)
+			}
+
+			if len(metas) >= limit {
+				return false
+			}
+
+			if page.IsTruncated == nil {
+				return false
+			}
+			return *page.IsTruncated
+		})
+	return metas, err
+}
+
+func (s *Store) put(ctx context.Context, key string, body io.Reader) error {
+	s.newUploader()
+
+	input := &s3manager.UploadInput{
+		Body:   body,
+		Bucket: &s.S3Bucket,
+		Key:    &key,
+	}
+
+	_, err := s.uploader.UploadWithContext(ctx, input)
+	return err
+}
+
+// get downloads a value from a key. Context can be canceled.
+// This is safe for multiple go routines.
+func (s *Store) get(ctx context.Context, key string) ([]byte, error) {
+	s.newDownloader()
+
+	input := &s3.GetObjectInput{
+		Bucket: &s.S3Bucket,
+		Key:    &key,
+	}
+
+	buf := make([]byte, 0, 16384) // pre-allocated 16KB for the s3 object.
+	w := aws.NewWriteAtBuffer(buf)
+	_, err := s.downloader.DownloadWithContext(ctx, w, input)
+	if err != nil {
+		return nil, err
+	}
+	buf = w.Bytes()
+	return buf, nil
+}
+
+func (s *Store) newUploader() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.uploader == nil {
+		s.uploader = s3manager.NewUploaderWithClient(s.svc)
+	}
+}
+
+func (s *Store) newDownloader() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.downloader == nil {
+		s.downloader = s3manager.NewDownloaderWithClient(s.svc)
+	}
+}
```

Comment:
Is there any need in lazy initialisation of `s.uploader` and `s.downloader`? They don't seem ever change and could be instantiated in `NewStore` making the code a bit simplier to follow, doesn't it?